test/test_torch.py

```diff
@@ -4342,6 +4342,85 @@ def test_diff(self, device, dtype):
             scalar = torch.tensor(2, device=device, dtype=dtype)
             torch.diff(scalar)
 
+    @onlyOnCPUAndCUDA
+    @dtypesIfCUDA(*set(torch.testing.get_all_math_dtypes('cuda')))
+    @dtypes(*set(torch.testing.get_all_math_dtypes('cpu')))
+    def test_gradient(self, device, dtype):
+        inputs = (
+            ((5,), [2.73], 0, 1),
+            ((5, 5), 2.0, 0, 1),
+            ((2, 2, 2), [1.17, 3.34, 1.73], [0, 1, 2], 1),
+            ((4, 4, 4), [(1., 3.0, 6., 10.), (1., 3.0, 4., 90.), (1., 2.5, 3.4, 7.)], [0, 1, 2], 1),
+            ((4, 4, 4), 1.7 , [0, 1, 2], 1),
+            ((4, 3, 4), 1.7 , [-2, -1], 1),
+            ((4, 3, 4, 5), [1.7, 3.93], [0, -1], 1),
+            ((4, 3, 4, 5), 1.7 , [1, 2], 1),
+            ((4, 3, 4, 5), 1.7 , [0, 1], 1),
+            ((4, 3, 4, 5), 1.7 , [2, -1], 1),
+            ((4, 2, 3, 5), [(1., 3.0, 6., 10.), (1., 3.0), 2.1], [0, 1, 2], 1),
+        )
+
+        if dtype == torch.uint8:
+            return
+
+        low_precision = dtype == torch.half or dtype == torch.bfloat16
+        rtol, atol = (0, 1e-2) if low_precision else (1e-5, 1e-5)
+
+        for size, spacing, axis, edge_order in inputs:
+            for contig in [True, False]:
+                t = make_tensor(size, device, dtype, discontiguous=not contig)
+                t_n = t.cpu().numpy()
+                if isinstance(spacing, list):
```

Comment:
Why special handling for when spacing is a list?

It might be simpler to write a `gradient_ref` wrapper around np.gradient that has the same signature as our torch version but computes using NumPy